cmd/promtool/tsdb_test.go

```diff
@@ -165,6 +192,71 @@ func sortLines(buf string) string {
 	return strings.Join(lines, "\n")
 }
 
+func TestTSDBDumpSeries(t *testing.T) {
+	storage := promqltest.LoadedStorage(t, `
+		load 1m
+			metric{foo="bar", baz="abc"} 1 2 3 4 5
+			heavy_metric{foo="bar"} 5 4 3 2 1
+			heavy_metric{foo="foo"} 5 4 3 2 1
+	`)
+
+	expectedArray := []string{
+		`{"__name__":"heavy_metric","foo":"bar"}
+`,
+		`{"__name__":"heavy_metric","foo":"foo"}
+`,
+		`{"__name__":"metric","baz":"abc","foo":"bar"}
+`,
+	}
+
+	tests := []struct {
+		name     string
+		mint     int64
+		maxt     int64
+		match    []string
+		expected string
+	}{
+		{
+			name:     "default match",
+			match:    []string{"{__name__=~'(?s:.*)'}"},
+			expected: strings.Join(expectedArray, ""),
+		},
+		{
+			name:     "same matcher twice",
+			match:    []string{"{foo=~'.+'}", "{foo=~'.+'}"},
+			expected: strings.Join(expectedArray, ""),
+		},
+		{
+			name:     "no duplication",
+			match:    []string{"{__name__=~'(?s:.*)'}", "{baz='abc'}"},
+			expected: strings.Join(expectedArray, ""),
+		},
+		{
+			name:     "well merged",
+			match:    []string{"{__name__='heavy_metric'}", "{baz='abc'}"},
+			expected: strings.Join(expectedArray, ""),
```

Comment:
This seems extremely complicated, to arrive at exactly the same result over and over again.

Reply:
😂  Copy the `TestTSDBDump` test case and modify it to the correct one